desktop/src/main.rs

```diff
@@ -112,373 +120,408 @@ fn trace_path(_opt: &Opt) -> Option<&Path> {
-        // Poll UI events
-        event_loop.run(move |event, _window_target, control_flow| {
-            // Allow KeyboardInput.modifiers (ModifiersChanged event not functional yet).
-            #[allow(deprecated)]
-            match event {
-                winit::event::Event::LoopDestroyed => {
-                    player.lock().unwrap().flush_shared_objects();
-                    return;
-                }
+    let absolute_path = selected
+        .canonicalize()
+        .unwrap_or_else(|_| selected.to_owned());
 
-                // Core loop
-                winit::event::Event::MainEventsCleared => {
-                    let new_time = Instant::now();
-                    let dt = new_time.duration_since(time).as_micros();
-                    if dt > 0 {
-                        time = new_time;
-                        let mut player_lock = player.lock().unwrap();
-                        player_lock.tick(dt as f64 / 1000.0);
-                        next_frame_time = new_time + player_lock.time_til_next_frame();
-                        if player_lock.needs_render() {
-                            window.request_redraw();
-                        }
-                    }
-                }
+    Ok(Some(load_movie_from_path(&absolute_path, opt)?))
+}
 
-                // Render
-                winit::event::Event::RedrawRequested(_) => {
-                    // Don't render when minimized to avoid potential swap chain errors in `wgpu`.
-                    if !minimized {
-                        player.lock().unwrap().render();
-                    }
+struct App {
+    #[allow(dead_code)]
+    opt: Opt,
+    window: Rc<Window>,
+    event_loop: EventLoop<RuffleEvent>,
+    executor: Arc<Mutex<GlutinAsyncExecutor>>,
+    player: Arc<Mutex<Player>>,
+    movie: Option<Arc<SwfMovie>>,
+}
+
+impl App {
+    fn new(opt: Opt) -> Result<Self, Box<dyn std::error::Error>> {
+        let movie = if let Some(path) = opt.input_path.to_owned() {
+            Some(load_movie_from_path(&path, &opt)?)
+        } else {
+            match load_from_file_dialog(&opt)? {
+                Some(movie) => Some(movie),
+                None => {
+                    shutdown(&Ok(()));
+                    std::process::exit(0);
                 }
+            }
+        };
+
+        let icon_bytes = include_bytes!("../assets/favicon-32.rgba");
+        let icon = Icon::from_rgba(icon_bytes.to_vec(), 32, 32)?;
+
+        let event_loop: EventLoop<RuffleEvent> = EventLoop::with_user_event();
+        let window = if let Some((movie, movie_url)) = &movie {
+            let filename = movie_url
+                .path_segments()
+                .and_then(|segments| segments.last())
+                .unwrap_or_else(|| movie_url.as_str());
+
+            let movie_size = LogicalSize::new(movie.width(), movie.height())
+                .cast::<f64>()
+                // .to_physical(window.scale_factor());
+                .to_physical(1.0); // TODO
```

Comment:
Why TODO here?

Reply:
I wonder whether the scale factor could be retrieved before creating a window, so it would be still possible to set it using `WindowBuilder`, but I didn't find a way after some searching in `winit`. Maybe you know?

Reply:
We'd have to know which monitor the window will appear on, because DPI can be different per monitor. I guess winit would default to the primary monitor?

https://osspial.github.io/winit/el2-win-joy/winit/event_loop/struct.EventLoop.html#method.get_primary_monitor
https://osspial.github.io/winit/el2-win-joy/winit/monitor/struct.MonitorHandle.html